src/node_api.cc

```diff
@@ -46,26 +47,34 @@ void node_napi_env__::CallFinalizer(napi_finalize cb, void* data, void* hint) {
 
 template <bool enforceUncaughtExceptionPolicy>
 void node_napi_env__::CallFinalizer(napi_finalize cb, void* data, void* hint) {
-  if (destructing) {
-    // we can not defer finalizers when the environment is being destructed.
-    v8::HandleScope handle_scope(isolate);
-    v8::Context::Scope context_scope(context());
-    CallbackIntoModule<enforceUncaughtExceptionPolicy>(
-        [&](napi_env env) { cb(env, data, hint); });
-    return;
+  v8::HandleScope handle_scope(isolate);
+  v8::Context::Scope context_scope(context());
+  CallbackIntoModule<enforceUncaughtExceptionPolicy>(
+      [&](napi_env env) { cb(env, data, hint); });
+}
+
+void node_napi_env__::EnqueueFinalizer(v8impl::RefTracker* finalizer) {
+  napi_env__::EnqueueFinalizer(finalizer);
+  if (!finalization_scheduled) {
+    finalization_scheduled = true;
+    Ref();
+    node_env()->SetImmediate([this](node::Environment* node_env) {
+      finalization_scheduled = false;
+      Unref();
+      DrainFinalizerQueue();
+    });
+  }
+}
+
+void node_napi_env__::DrainFinalizerQueue() {
+  // As userland code can delete additional references in one finalizer,
+  // the list of pending finalizers may be mutated as we execute them, so
+  // we keep iterating it until it is empty.
+  while (!pending_finalizers.empty()) {
+    v8impl::RefTracker* ref_tracker = *pending_finalizers.begin();
+    pending_finalizers.erase(ref_tracker);
+    ref_tracker->Finalize();
```

Comment:
How do we handle uncaught JS exceptions? Previously the `SetImmediate` code was taking care about the exceptions. It may be worth adding a test where we have a few finalizers running and they all throw JS exceptions. The JS code should handle and ignore those exceptions - we must see all the finalizers executed.
Also, we need to test that if JS does not handle the uncaught exceptions, then we crash as usual.

Reply:
`node_napi_env__::CallFinalizer` still handles JS exception. Native `SetImmediate` doesn't handle JS exceptions.